mail_base/static/src/js/base.js

```diff
@@ -0,0 +1,1083 @@
+        if ((chat_session.state === "open") || (chat_session.state === "folded")) {
+            channel = chat_session.is_minimized && chat_manager.get_channel(chat_session.id);
+            if (channel) {
+                channel.is_detached = true;
+                channel.is_folded = (chat_session.state === "folded");
+                chat_manager.bus.trigger("open_chat", channel);
+            }
+        } else if (chat_session.state === "closed") {
+            channel = chat_manager.get_channel(chat_session.id);
+            if (channel) {
+                channel.is_detached = false;
+                chat_manager.bus.trigger("close_chat", channel, {keep_open_if_unread: true});
+            }
+        }
+    },
+
+    on_presence_notification: function (data) {
+        var dm = chat_manager.get_dm_from_partner_id(data.id);
+        if (dm) {
+            dm.status = data.im_status;
+            chat_manager.bus.trigger('update_dm_presence', dm);
+        }
+    },
+
+    start: function(){
+        this.add_channel({
+            id: "channel_inbox",
+            name: _t("Inbox"),
+            type: "static"
+        }, { display_needactions: true });
+
+        this.add_channel({
+            id: "channel_starred",
+            name: _t("Starred"),
+            type: "static"
+        });
+
+        var load_channels = session.rpc('/mail/client_action').then(function (result) {
+            _.each(result.channel_slots, function (channels) {
+                _.each(channels, cls.add_channel);
+            });
+            needaction_counter = result.needaction_inbox_counter;
+            mention_partner_suggestions = result.mention_partner_suggestions;
+        });
+
+        var load_emojis = session.rpc("/mail/chat_init").then(function (result) {
+            emojis = result.emoji;
+            _.each(emojis, function(emoji) {
+                emoji_substitutions[_.escape(emoji.source)] = emoji.substitution;
+            });
+        });
+
+        var ir_model = new Model("ir.model.data");
+        var load_menu_id = ir_model.call("xmlid_to_res_id", ["mail.mail_channel_menu_root_chat"], {}, {shadow: true});
+        var load_action_id = ir_model.call("xmlid_to_res_id", ["mail.mail_channel_action_client_chat"], {}, {shadow: true});
+
+        bus.on('notification', null, cls.on_notification);
+
+        return $.when(load_menu_id, load_action_id, load_channels, load_emojis).then(function (menu_id, action_id) {
+            discuss_ids = {
+                menu_id: menu_id,
+                action_id: action_id
+            };
+            bus.start_polling();
+        });
+    }
+});
+
+var cls = new MailTools();
+
+// Public interface
+//----------------------------------------------------------------------------------
+// these two functions are exposed for extensibility purposes and shouldn't be called by other modules
+chat_manager.make_message = cls.make_message;
+chat_manager.make_channel = cls.make_channel;
```

Comment:
add a comment

     // we add this function this way in order to make them extendable via MailTools.include({...})